warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/job/ShardReindexJob.java

```diff
@@ -0,0 +1,522 @@
+        // Key end = new Key("z", "fi~");
+        // Range startRange = new Range(start, true, end, true);
+        //
+        // Range subRange = new Range(new Key("c"), true, new Key("d"), true);
+        //
+        // Range clipped = subRange.clip(startRange);
+        //
+        // System.out.println(clipped.getStartKey());
+        //
+        // return;
+
+        System.exit(ToolRunner.run(null, new ShardReindexJob(), args));
+    }
+
+    public static class FiToGiMapper extends Mapper<Key,Value,BulkIngestKey,Value> {
+        private static final Logger log = Logger.getLogger(FiToGiMapper.class);
+        private final byte[] FI_START_BYTES = FI_START.getBytes();
+        private final Value UID_VALUE = new Value(buildIndexValue().toByteArray());
+        private final Value EMPTY_VALUE = new Value();
+
+        private TypeRegistry typeRegistry;
+        private Map<String,IngestHelperInterface> datatypeHelperCache;
+        private boolean cleanupShard;
+
+        private Text shardTable;
+        private Text indexTable;
+        private Text reverseIndexTable;
+
+        private byte[] lastFiBytes;
+        private String field;
+
+        private boolean propagateDeletes;
+
+        @Override
+        protected void setup(Context context) throws IOException, InterruptedException {
+            Configuration config = context.getConfiguration();
+            this.typeRegistry = TypeRegistry.getInstance(config);
+
+            this.cleanupShard = config.getBoolean("job.cleanupShard", false);
+
+            this.shardTable = new Text(config.get(ShardedDataTypeHandler.SHARD_TNAME, "shard"));
+            this.indexTable = new Text(config.get(ShardedDataTypeHandler.SHARD_GIDX_TNAME, "shardIndex"));
+            this.reverseIndexTable = new Text(config.get(ShardedDataTypeHandler.SHARD_GRIDX_TNAME, "shardReverseIndex"));
+
+            this.propagateDeletes = config.getBoolean("propagateDeletes", false);
+
+            this.datatypeHelperCache = new HashMap<>();
+        }
+
+        @Override
+        protected void map(Key key, Value value, Context context) throws IOException, InterruptedException {
+            // This is all a bit awkward since DatawaveKey already parses the Key, but is in the datawave-query-core package. Minimally parse the Key for three
+            // purposes
+            // 1. Is it an FI key. Ranges should be created to defeat non fi ranges, but no verify and skip when possible
+            // 2. Get the field name. Necessary for checking how a field is indexed
+            // 3. Get the data type. The data type is used to get the correct IngestHelperInterface which is used to make decisions regarding how a field is
+            // indexed. This may vary from data type to data type
+
+            // ensure the key is an fi
+            final byte[] cf = key.getColumnFamilyData().getBackingArray();
+            if (WritableComparator.compareBytes(cf, 0, 3, FI_START_BYTES, 0, FI_START_BYTES.length) != 0) {
```

Comment:
Need to check the lenth of cf first.  There will be keys in the shard table where the cf is less than 3 and this will throw an exception.

Reply:
Also shouldn't that first 3 also be FI_START_BYTES.length ?